crates/astria-core/src/sequencerblock/v1/block/mod.rs

```diff
@@ -1165,122 +1162,424 @@ impl ExtendedCommitInfoError {
-    let rollup_ids_proof = tree.construct_proof(1).expect(
-        "the leaf must exist in the tree as `rollup_ids_root` was created from the same index in \
-         `data` used to construct the tree",
-    );
-
-    let (extended_commit_info, extended_commit_info_proof) = if with_extended_commit_info {
-        let extended_commit_info = data_iter
+#[derive(Debug, thiserror::Error)]
+enum DataItemErrorKind {
+    #[error("failed to borsh-encode {variant} as data item")]
+    Encode {
+        source: io::Error,
+        variant: &'static str,
+    },
+    #[error("item {index} of cometbft `block.data` could not be borsh-decoded")]
+    Decode { source: io::Error, index: usize },
+    #[error("unexpected data element `{actual}` at index {index}, expected `{expected}`")]
+    Mismatch {
+        index: usize,
+        expected: &'static str,
+        actual: &'static str,
+    },
+}
+
+/// The parsed elements of the `data` field of a `SequencerBlock`, along with values derived from
+/// these.
+#[derive(Debug)]
+pub struct ParsedDataItems {
+    /// The root hash of the Merkle tree derived from the `data` (not itself an actual entry in
+    /// `data`).
+    pub data_root_hash: [u8; SHA256_DIGEST_LENGTH],
+    /// The first item in `data`: the root hash of the Merkle tree derived from the rollup txs.
+    pub rollup_transactions_root: [u8; SHA256_DIGEST_LENGTH],
+    /// The proof that `rollup_transactions_root` exists in the data Merkle tree (not itself an
+    /// actual entry in `data`).
+    pub rollup_transactions_proof: merkle::Proof,
+    /// The second item in `data`: the root hash of the Merkle tree derived from the rollup IDs.
+    pub rollup_ids_root: [u8; SHA256_DIGEST_LENGTH],
+    /// The proof that `rollup_ids_root` exists in the data Merkle tree (not itself an actual entry
+    /// in `data`).
+    pub rollup_ids_proof: merkle::Proof,
+    /// The optional third item in `data`: the collection of `ChangeHash`es applied as part of an
+    /// upgrade executed during this block, along with the proof that this entry exists in the data
+    /// Merkle tree. If no such upgrade exists, there will be no entry in `data` for this, as
+    /// opposed to an empty or `None` entry.
+    pub upgrade_change_hashes_with_proof: Option<UpgradeChangeHashesWithProof>,
+    /// The optional next item in `data` (third or fourth depending on whether an upgrade exists in
+    /// `data`): the encoded extended commit info, along with the proof that this entry exists in
+    /// the data Merkle tree. If extended commit info is enabled at this block height, this entry
+    /// must exist. If not, there must be no entry in `data` for this, as opposed to an empty or
+    /// `None` entry.
+    pub extended_commit_info_with_proof: Option<ExtendedCommitInfoWithProof>,
+    /// All remaining entries of `data`, parsed into rollup txs, in the same order as they appear
+    /// in `data`.
+    pub rollup_transactions: Vec<Arc<Transaction>>,
```

Comment:
```suggestion
    /// All remaining entries of `data`, parsed into txs, in the same order as they appear
    /// in `data`.
    pub transactions: Vec<Arc<Transaction>>,
```
naming is confusing, as not all user-submitted txs are for rollups

Reply:
Renamed to `user_submitted_transactions` in [4a045d3](https://github.com/astriaorg/astria/pull/1917/commits/4a045d36c177676c3af883aed37da3cce61f30b6).